### ISSUE TITLE
Idea: Detect out-of-date environment in ”shell” and “run”

### ISSUE THREAD

**Issue by uranusjr** [pypa/pipenv#1924]
Was working on Pipenv and got an `ImportError` when I do `pipenv run ...`. Took me a minute to realise “oh yeah there’s a new dependency, I need to `pipenv sync`”.

It would be pretty neat if Pipenv can tell me something like

```
$ pipenv run ...
Notice: Your environment is out of sync to Pipfile.lock. Run pipenv sync or pipenv sync --dev to update!
[Attempt to run the script anyway]
```

I’m not sure if this is even feasible, but it would be a very-nice-to-have.

**Comment by techalchemy** [comment]
It should be

**Comment by ncoghlan** [comment]
We wouldn't be scanning the import names, we'd be comparing the PEP 376 "installed distribution packages" info with the pinned distribution package names in the lock file.

This would be akin to what `pip-sync` does in `pip-tools`, but without automatically fixing any discrepancies.

However, I'm still not sure we'd want to do this, as I'd expect the runtime cost to be pretty high, and it would scale with both `Pipfile.lock` size and `sys.path` length.

**Comment by uranusjr** [comment]
I agree the runtime cost would be (too) high. Maybe it’s enough to hijack pip’s message and display something else? Replacing the upgrade suggestion with `pipenv run python -m pip install --upgrade pip`, for example. Pipenv already does this message replacement somewhere, iirc.

**Comment by jkleint** [comment]
+1.  I'd like to see this for `pipenv shell`, as you only do that once so the overhead isn't such an issue.

**Comment by uranusjr** [comment]
It still could take a while though. Maybe it’s a good idea to implement it as an opt-in, but that would make it a low-priority feature. I’m open to contributions if anyone is interested, however.

**Comment by ncoghlan** [comment]
Something that we could potentially do to make this a relatively fast check is to add a metadata file inside the `venv` directory that captures the lock file hash (which would thus go stale when `git pull` replaced your lock file).

I'd still be wary of adding that check to `pipenv run` for startup latency reasons, but that concern could be handled via a `--skip-sync-check` flag.
